### PR TITLE
OSU MPI Microbenchmarks for MVAPICH MPI testcase

### PR DESCRIPTION
**Pull request by ekmalkan**
Copy of PR : https://github.com/LIS/LISAv2/pull/797

Added new test case for OSU Benchmarks for MVAPICH MPI.
Ran the test locally and received PASS for multiple trials.

[LISAv2 Test Results Summary]
Test Run On           : 05/20/2020 01:56:17
ARM Image Under Test  : OpenLogic : CentOS : 7.7 : latest
Initial Kernel Version: 3.10.0-1062.12.1.el7.x86_64
Final Kernel Version  : 3.10.0-1062.12.1.el7.x86_64
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:19

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 INFINIBAND           INFINIBAND-MVAPICH-MPI-OMB-2VM                                                    PASS                15.16
        InfiniBand-Verification-1-FirstBoot : eth0 IP : PASS
        InfiniBand-Verification-1-FirstBoot : IBV_PINGPONG : PASS
        InfiniBand-Verification-1-FirstBoot : PingPong Intranode : PASS
        InfiniBand-Verification-1-FirstBoot : IMB-MPI1 : PASS
        InfiniBand-Verification-1-FirstBoot : OMB-P2P : PASS
        InfiniBand-Verification-2-Reboot : eth0 IP : PASS
        InfiniBand-Verification-2-Reboot : IBV_PINGPONG : PASS
        InfiniBand-Verification-2-Reboot : PingPong Intranode : PASS
        InfiniBand-Verification-2-Reboot : IMB-MPI1 : PASS
        InfiniBand-Verification-2-Reboot : OMB-P2P : PASS

Logs can be found at C:\Users\ekmalkan\LisaV2Project\TestResults\2020-19-05-18-56-09-4115

### CODE REVIEW

**Comment by ekmalkan** [comment]
[Azure-UD64-TestLogs.zip](https://github.com/LIS/LISAv2/files/4654288/Azure-UD64-TestLogs.zip)

**Comment by LiliDeng** [comment]
@ekmalkan test case abort against canonical ubuntuserver 18.04-lts Latest, is it expected?

**Comment by ekmalkan** [comment]
This seems to be passing with canonical ubuntuserver 18.04-lts Latest on running locally. Test logs attached
[Azure-QL13-TestLogs.zip](https://github.com/LIS/LISAv2/files/4659124/Azure-QL13-TestLogs.zip)

Can you elaborate on how you ran the test/ query used etc?

File: Testscripts/Linux/SetupRDMA.sh [hunk 1]
```diff
@@ -613,8 +613,29 @@ function Main() {
 		Verify_File $benchmark_bin
 	fi
 
+	if [ $mpi_type != "mvapich" ]; then
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SetupRDMA.sh]: you don't need this if-statement, it should be under "# MVAPICH MPI installation" else-block.

File: Testscripts/Linux/SetupRDMA.sh [hunk 2]
```diff
@@ -613,8 +613,29 @@ function Main() {
...
+		LogMsg "Downloading mpi-benchmarks from $osu_mpi_benchmark"
+		wget $osu_mpi_benchmark
+		tar_filename=$(echo $osu_mpi_benchmark | rev | cut -d'/' -f1 | rev)
+		tar xvzf tar_filename
+		LogMsg "Untarred $tar_filename"
+		cd ${tar_filename%.*.*}
+
+		LogMsg "Running configuration and installing"
+		./configure CC=/usr/local/bin/mpicc CXX=/usr/local/bin/mpicxx --prefix=$(pwd) && make && make install
+		Verify_Result
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/SetupRDMA.sh:635]: This Verify_Result only works for the last command, 'make install'

File: Testscripts/Linux/TestRDMA_MultiVM.sh [hunk 3]
```diff
@@ -169,6 +174,43 @@ function Run_IMB_MPI1() {
 	fi
 }
 
+function Run_OMB_P2P() {
+	total_attempts=$(seq 1 1 $omb_p2p_tests_iterations)
+	omb_p2p_tests_array=($omb_p2p_tests)
+	for attempt in $total_attempts; do
+		LogMsg "OMB P2P test iteration $attempt for $mpi_type- Running."
+		case "$mpi_type" in
+			ibm | open | hpcx | intel)
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/TestRDMA_MultiVM.sh]: Please them in 'default' and move 'mvapich' to the first block.

File: Testscripts/Linux/TestRDMA_MultiVM.sh [hunk 4]
```diff
@@ -169,6 +174,43 @@ function Run_IMB_MPI1() {
 	fi
 }
 
+function Run_OMB_P2P() {
+	total_attempts=$(seq 1 1 $omb_p2p_tests_iterations)
+	omb_p2p_tests_array=($omb_p2p_tests)
+	for attempt in $total_attempts; do
+		LogMsg "OMB P2P test iteration $attempt for $mpi_type- Running."
```
Comment (juhlee-microsoft) [inline Testscripts/Linux/TestRDMA_MultiVM.sh]: This log should be placed in the line 187 block. If it runs with non-mvapich mode, the logs do not make sense.

File: XML/TestCases/BenchmarkTests-MPI.xml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		<TestParameters>
+			<param>install_ofed="INFINIBAND_INSTALL_OFED"</param>
+			<param>install_ofed_from_extension="INFINIBAND_INSTALL_OFED_FROM_EXTENSION"</param>
+			<param>num_reboot="INFINIBAND_MVAPICH_TOTAL_REBOOT_COUNT"</param>
+			<param>mpi_settings="INFINIBAND_OMB_MVAPICH_MPI_SETTINGS"</param>
+			<param>ib_nic="INFINIBAND_IB_NIC"</param>
+			<param>imb_mpi1_tests="INFINIBAND_MVAPICH_MPI_IMB-TESTNAMES"</param>
+			<param>mpi1_ppn=INFINIBAND_MVAPICH_PPN</param>
+			<param>imb_mpi1_tests_iterations=INFINIBAND_MPI_TEST_COUNT</param>
+            <param>omb_p2p_tests="INFINIBAND_MVAPICH_OMB_P2P_TESTNAMES"</param>
```
Comment (juhlee-microsoft) [inline XML/TestCases/BenchmarkTests-MPI.xml]: Please check out indentation for the next 4 lines.

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/SetupRDMA.sh (hunk 1 above)
Comment (ekmalkan) [reply Testscripts/Linux/SetupRDMA.sh]: Added another check in the if statement to install this for OMB benchmarks tests (excluding MVAPICH as it has omb preinstalled)

File: Testscripts/Linux/SetupRDMA.sh (hunk 2 above)
Comment (ekmalkan) [reply Testscripts/Linux/SetupRDMA.sh:635]: Broke down this cmmand and added Verify_Result for each subcommand

File: Testscripts/Linux/TestRDMA_MultiVM.sh (hunk 4 above)
Comment (ekmalkan) [reply Testscripts/Linux/TestRDMA_MultiVM.sh]: done in revision

File: XML/TestCases/BenchmarkTests-MPI.xml (hunk 5 above)
Comment (ekmalkan) [reply XML/TestCases/BenchmarkTests-MPI.xml]: done in revision. strangely the indentation shows correctly in vs code and not in commit sometimes.

File: Testscripts/Linux/TestRDMA_MultiVM.sh (hunk 3 above)
Comment (ekmalkan) [reply Testscripts/Linux/TestRDMA_MultiVM.sh]: done in revision

**Comment by ekmalkan** [comment]
[Azure-RL84-TestLogs.zip](https://github.com/LIS/LISAv2/files/4659727/Azure-RL84-TestLogs.zip)

Test logs attached (post revision)

**Comment by juhlee-microsoft** [comment]
> This seems to be passing with canonical ubuntuserver 18.04-lts Latest on running locally. Test logs attached
> [Azure-QL13-TestLogs.zip](https://github.com/LIS/LISAv2/files/4659124/Azure-QL13-TestLogs.zip)
> 
> Can you elaborate on how you ran the test/ query used etc?

@ekmalkan Can you please test without QUICKTEST_ONLY=yes? This is for kernel testing, and this parameter does not exercise all paths.

**Comment by ekmalkan** [comment]
This seems to be passing with canonical ubuntuserver 18.04-lts Latest on running locally. Test logs attached
Azure-QL13-TestLogs.zip

Can you elaborate on how you ran the test/ query used etc?

@ekmalkan Can you please test without QUICKTEST_ONLY=yes? This is for kernel testing, and this parameter does not exercise all paths.

@juhlee-microsoft ,  QUICKTEST_ONLY=yes should ideally apply only for IMB benchmark tests and not the new OMB benchmarks. I have pushed changes to make the seperation more clear. With the new changes,  QUICKTEST_ONLY parameter does not apply to OMB tests

**Comment by ekmalkan** [comment]
Test logs 
1.OMB MVAPICH 2VM,
[Azure-NY40-TestLogs.zip](https://github.com/LIS/LISAv2/files/4669999/Azure-NY40-TestLogs.zip)
[LISAv2 Test Results Summary]
Test Run On           : 05/22/2020 16:23:50
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Initial Kernel Version: 5.3.0-1020-azure
Final Kernel Version  : 5.3.0-1022-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:24

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 INFINIBAND           INFINIBAND-MVAPICH-MPI-OMB-2VM                                                    PASS                17.62 
	InfiniBand-Verification-1-FirstBoot : eth0 IP : PASS 
	InfiniBand-Verification-1-FirstBoot : IBV_PINGPONG : PASS 
	InfiniBand-Verification-1-FirstBoot : OMB-P2P : PASS 
	InfiniBand-Verification-2-Reboot : eth0 IP : PASS 
	InfiniBand-Verification-2-Reboot : IBV_PINGPONG : PASS 
	InfiniBand-Verification-2-Reboot : OMB-P2P : PASS 


Logs can be found at C:\Users\ekmalkan\LisaV2Project\TestResults\2020-22-05-09-23-43-2951

**Comment by ekmalkan** [comment]
2. IMB MVAPICH 2VM (full test i.e.  QUICKTEST_ONLY=no)

[LISAv2 Test Results Summary]
Test Run On           : 05/22/2020 00:51:29
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Initial Kernel Version: 5.3.0-1020-azure
Final Kernel Version  : 5.3.0-1022-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:4:27

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 INFINIBAND           INFINIBAND-MVAPICH-MPI-2VM                                                        PASS               261.51 
	InfiniBand-Verification-1-FirstBoot : eth0 IP : PASS 
	InfiniBand-Verification-1-FirstBoot : IBV_PINGPONG : PASS 
	InfiniBand-Verification-1-FirstBoot :IMB PingPong Intranode : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-MPI1 : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-P2P : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-IO : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-RMA : PASS 
	InfiniBand-Verification-1-FirstBoot : IMB-NBC : PASS 
	InfiniBand-Verification-2-Reboot : eth0 IP : PASS 
	InfiniBand-Verification-2-Reboot : IBV_PINGPONG : PASS 
	InfiniBand-Verification-2-Reboot :IMB PingPong Intranode : PASS 
	InfiniBand-Verification-2-Reboot : IMB-MPI1 : PASS 
	InfiniBand-Verification-2-Reboot : IMB-P2P : PASS 
	InfiniBand-Verification-2-Reboot : IMB-IO : PASS 
	InfiniBand-Verification-2-Reboot : IMB-RMA : PASS 
	InfiniBand-Verification-2-Reboot : IMB-NBC : PASS 


Logs can be found at C:\Users\ekmalkan\LisaV2Project\TestResults\2020-21-05-17-51-22-7485

Logs too huge to post here

File: XML/TestCases/BenchmarkTests-MPI.xml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<TestCases>
+<!-- MVAPICH MPI -->
+    <test>
+		<testName>INFINIBAND-MVAPICH-MPI-OMB-2VM</testName>
```
Comment (juhlee-microsoft) [inline XML/TestCases/BenchmarkTests-MPI.xml:4]: Can you check here for indentation? I am not sure if you have double tabs or Github UI issue. My question is all for other change lines. It looks mixed, so I guess you mixed tab and space or different size. LISAv2 has tab, size 4 in the how to guideline. Please make sure of your indentation on all changes.

File: XML/Other/ReplaceableTestParameters.xml [hunk 7]
```diff
@@ -228,6 +232,10 @@ Scenario 2 : You need to run all the performance tests quickly.
 	<Parameter>
 		<ReplaceThis>INFINIBAND_MVAPICH_PPN</ReplaceThis>
 		<ReplaceWith>1</ReplaceWith>
+	</Parameter>
+		<Parameter>
```
Comment (juhlee-microsoft) [inline XML/Other/ReplaceableTestParameters.xml]: indentation?

**Comment by ekmalkan** [comment]
[LISAv2 Test Results Summary]
Test Run On           : 05/22/2020 22:53:12
ARM Image Under Test  : canonical : ubuntuserver : 18.04-lts : Latest
Initial Kernel Version: 5.3.0-1020-azure
Final Kernel Version  : 5.3.0-1022-azure
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:34

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes)
-------------------------------------------------------------------------------------------------------------------------------------------
    1 INFINIBAND           INFINIBAND-MVAPICH-MPI-OMB-2VM                                                    PASS                29.76
        InfiniBand-Verification-1-FirstBoot : eth0 IP : PASS
        InfiniBand-Verification-1-FirstBoot : IBV_PINGPONG : PASS
        InfiniBand-Verification-1-FirstBoot : OMB-P2P : PASS
        InfiniBand-Verification-2-Reboot : eth0 IP : PASS
        InfiniBand-Verification-2-Reboot : IBV_PINGPONG : PASS
        InfiniBand-Verification-2-Reboot : OMB-P2P : PASS

[Azure-SK24-TestLogs.zip](https://github.com/LIS/LISAv2/files/4670744/Azure-SK24-TestLogs.zip)

File: XML/Other/ReplaceableTestParameters.xml (hunk 7 above)
Comment (ekmalkan) [reply XML/Other/ReplaceableTestParameters.xml]: done

**Review by juhlee-microsoft — APPROVED** [review]

**Pull request merged**